File: boxblur.lua
--box blur algorithm by Mario Klingemann http://incubator.quasimondo.com
--src and dst are two image objects per bmpconv
local ffi = require'ffi'

local function rgb(pix, i)
	local p = pix[i]
	return
		bit.rshift(bit.band(p, 0xff0000), 16),
		bit.rshift(bit.band(p, 0x00ff00), 8),
		bit.band(p, 0x0000ff)
end

local function irgb(r, g, b)
	return bit.bor(0xff000000, bit.lshift(r, 16), bit.lshift(g, 8), b)
end

local function boxblur(src, radius)
	if radius < 1 then return end
	local w, h = src.w, src.h
	local div = 2*radius+1
	local r = ffi.new('uint8_t[?]', w*h)
	local g = ffi.new('uint8_t[?]', w*h)
	local b = ffi.new('uint8_t[?]', w*h)
	local vmin = ffi.new('int32_t[?]', math.max(w, h))
	local vmax = ffi.new('int32_t[?]', math.max(w, h))
	local pix = ffi.cast('int32_t*', src.data)
	local dv = ffi.new('uint8_t[?]', 256*div)
	for i=0,256*div-1 do dv[i] = i/div end

	local yw, yi = 0, 0

	for x=0,w-1 do
		vmin[x] = math.min(x+radius+1, w-1)
		vmax[x] = math.max(x-radius, 0)
	end

	for y=0,h-1 do
		local rsum, gsum, bsum = 0, 0, 0
		for i=-radius,radius do
			local rr, gg, bb = rgb(pix, yi+math.min(w-1, math.max(i, 0)))
			rsum = rsum + rr
			gsum = gsum + gg
			bsum = bsum + bb
		end
		for x=0,w-1 do
			r[yi] = dv[rsum]
			g[yi] = dv[gsum]
			b[yi] = dv[bsum]
			local r1,g1,b1 = rgb(pix, yw+vmin[x])
			local r2,g2,b2 = rgb(pix, yw+vmax[x])
			rsum = rsum + r1-r2
			gsum = gsum + g1-g2
			bsum = bsum + b1-b2
			yi = yi + 1
		end
		yw = yw+w
	end

	for y=0,h-1 do
		vmin[y] = math.min(y+radius+1, h-1) * w
		vmax[y] = math.max(y-radius, 0) * w
	end

	for x=0,w-1 do
		local rsum, gsum, bsum = 0, 0, 0
		local yp = -radius * w
		for i=-radius,radius do
			yi = math.max(0, yp) + x
			rsum = rsum + r[yi]
			gsum = gsum + g[yi]
			bsum = bsum + b[yi]
			yp = yp + w
		end

		yi = x
		for y=0,h-1 do
			pix[yi] = irgb(dv[rsum], dv[gsum], dv[bsum])
			local p1 = x+vmin[y]
			local p2 = x+vmax[y]
			rsum = rsum + r[p1]-r[p2]
			gsum = gsum + g[p1]-g[p2]
			bsum = bsum + b[p1]-b[p2]
			yi = yi + w
		end
	end
end

if not ... then require'test' end

return boxblur

File: csrc/boxblur/boxblur.c
//go@ gcc boxblur.c -O3 -o ../../bin/boxblur.dll -shared
//Box Blur Algorithm by Mario Klingemann http://incubator.quasimondo.com

#include <stdint.h>
#include <stdlib.h>
#include "branchless.c"

void box_blur_argb32(uint8_t *pix, int32_t w, int32_t h, int32_t radius) {
	if (radius < 1) return;

	int32_t rsum, gsum, bsum, x, y, i, p, p1, p2, yp, yi, yw;

	uint8_t* r = malloc(w * h);
	uint8_t* g = malloc(w * h);
	uint8_t* b = malloc(w * h);
	int32_t *vmin = malloc(max(w,h) * 4);
	int32_t *vmax = malloc(max(w,h) * 4);

	int32_t div = 2*radius+1;
	uint8_t* dv = malloc(256*div);
	for (i=0; i<256*div; i++)
		dv[i] = (i/div);

	yw = yi = 0;

	for (x=0; x<w; x++) {
		vmin[x] = min(x+radius+1,w-1);
		vmax[x] = max(x-radius,0);
	}

	for (y=0; y<h; y++) {
		rsum = gsum = bsum = 0;

		for(i=-radius; i<=radius; i++) {
			p = (yi + min(w-1, max(i,0))) * 4;
			rsum += pix[p];
			gsum += pix[p+1];
			bsum += pix[p+2];
		}

		for (x=0; x<w; x++) {
			r[yi] = dv[rsum];
			g[yi] = dv[gsum];
			b[yi] = dv[bsum];

			p1 = (yw+vmin[x])*4;
			p2 = (yw+vmax[x])*4;

			rsum += pix[p1] - pix[p2];
			gsum += pix[p1+1] - pix[p2+1];
			bsum += pix[p1+2] - pix[p2+2];

			yi++;
		}
		yw += w;
	}

	for (y=0; y<h; y++) {
		vmin[y] = min(y+radius+1,h-1)*w;
		vmax[y] = max(y-radius,0)*w;
	}

	for (x=0; x<w; x++) {
		rsum = gsum = bsum = 0;
		yp = -radius*w;

		for(i=-radius; i<=radius; i++) {
			yi = max(0,yp)+x;
			rsum += r[yi];
			gsum += g[yi];
			bsum += b[yi];
			yp += w;
		}
		yi = x;

		for (y=0; y<h; y++) {
			pix[yi*4 + 0] = dv[rsum];
			pix[yi*4 + 1] = dv[gsum];
			pix[yi*4 + 2] = dv[bsum];

			p1 = x+vmin[y];
			p2 = x+vmax[y];

			rsum += r[p1]-r[p2];
			gsum += g[p1]-g[p2];
			bsum += b[p1]-b[p2];

			yi += w;
		}
	}

	free(r);
	free(g);
	free(b);
	free(vmin);
	free(vmax);
	free(dv);
}
